fv3core/stencils/delnflux.py

```diff
@@ -1,218 +1,627 @@
         )
-    return fx, fy
-
-
-def compute_no_sg(q, fx2, fy2, nord, damp_c, d2, kstart=0, nk=None, mass=None):
-    # q is input only
-    grid = spec.grid
-    nord = int(nord)
-    i1 = grid.is_ - 1 - nord
-    i2 = grid.ie + 1 + nord
-    j1 = grid.js - 1 - nord
-    j2 = grid.je + 1 + nord
-    if nk is None:
-        nk = grid.npz - kstart
-    kslice = slice(kstart, kstart + nk)
-    origin_d2 = (i1, j1, kstart)
-    domain_d2 = (i2 - i1 + 1, j2 - j1 + 1, nk)
-    if mass is None:
-        d2_damp(q, d2, damp_c, origin=origin_d2, domain=domain_d2)
-    else:
-        copy_stencil(q, d2, origin=origin_d2, domain=domain_d2)
-
-    if nord > 0:
-        corners.copy_corners_x_stencil(
-            d2, origin=(grid.isd, grid.jsd, kstart), domain=(grid.nid, grid.njd, nk)
+
+        self._damping_factor_calculation(self._damp_3d, nord, damp_c, self._grid.da_min)
+        self._damp = utils.make_storage_data(self._damp_3d[0, 0, :], (nk,), (0,))
+
+        self.delnflux_nosg = DelnFluxNoSG(nord, nk=nk)
+
+    def __call__(
+        self,
+        q: FloatField,
+        fx: FloatField,
+        fy: FloatField,
+        d2: Optional["FloatField"] = None,
+        mass: Optional["FloatField"] = None,
+    ):
+        """
+        Del-n damping for fluxes, where n = 2 * nord + 2
+        Args:
+            q: Field for which to calculate damped fluxes (in)
+            fx: x-flux on A-grid (inout)
+            fy: y-flux on A-grid (inout)
+            d2: A damped copy of the q field (in)
+            mass: Mass to weight the diffusive flux by (in)
+        """
+        if self._no_compute is True:
+            return fx, fy
+
+        if d2 is None:
+            d2 = utils.make_storage_from_shape(
+                q.shape, self._origin, cache_key="delnflux_d2"
+            )
+
+        self.delnflux_nosg(q, self._fx2, self._fy2, self._damp, d2, mass)
+
+        if mass is None:
+            self._add_diffusive_stencil(fx, self._fx2, fy, self._fy2)
+        else:
+            # TODO: To join these stencils you need to overcompute, making the edges
+            # 'wrong', but not actually used, separating now for comparison sanity.
+
+            # diffusive_damp(fx, fx2, fy, fy2, mass, damp, origin=diffuse_origin,
+            # domain=(grid.nic + 1, grid.njc + 1, nk))
+            self._diffusive_damp_stencil(fx, self._fx2, fy, self._fy2, mass, self._damp)
+
+        return fx, fy
+
+
+class DelnFluxNoSG:
+    """
+    This contains the mechanics of del6_vt and some of deln_flux from
+    the Fortran code, since they are very similar routines. The test class
+    is Del6VtFlux
+    """
+
+    def __init__(self, nord, nk: Optional[int] = None):
+        """
+        nord sets the order of damping to apply:
+        nord = 0:   del-2
+        nord = 1:   del-4
+        nord = 2:   del-6
+        """
+        if max(nord[:]) > 3:
+            raise Exception("nord must be less than 3")
```

Comment:
Fair point, though it's not quite a "NotImplemented" issue because nord > 3 is not actually possible with the FV3 configuration, as it'd require larger halos